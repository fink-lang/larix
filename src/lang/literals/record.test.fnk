{describe, it, expect, to_match_snapshot} = import '@fink/jest/test.fnk'

{parse_expr, parse} = import '../../parser.fnk'



describe 'record {...}', fn:

  it 'parses empty: {}', fn:
    expect
      parse_expr '{}'
      to_match_snapshot


  it 'parses single shorthand prop', fn:
    expect
      parse_expr '{foo}'

      to_match_snapshot


  it 'parses single shorthand value', fn:
    expect
      parse_expr "{'foo'}"
      to_match_snapshot


  it 'parses multiple shorthand exprs', fn:
    expect
      parse_expr '{foo, bar, π}'
      to_match_snapshot


  it 'parses multiple shorthand exprs', fn:
    expect
      parse_expr "{foo, bar: 123, shrub: 'ni'}"
      to_match_snapshot


  it 'parses single prop', fn:
    expect
      parse_expr '{foo: spam}'
      to_match_snapshot


  it 'parses spread prop', fn:
    expect
      parse_expr '{foo, ...bar}'
      to_match_snapshot


  it 'parses single str prop}', fn:
    expect
      parse_expr "{'foo': spam}"
      to_match_snapshot


  it 'parses default assignment prop', fn:
    expect
      parse_expr '{foo=123}'
      to_match_snapshot


  it 'parses calculated prop', fn:
    expect
      parse_expr '{foo, (bar + ni): 123, spam}'
      to_match_snapshot


  it 'parses keywords as keys', fn:
    expect
      parse_expr '
        {fn, match, pipe, rec, list, await, throw, try}
      '
      to_match_snapshot

    expect
      parse_expr '
        rec:
          fn: 123
          rec: 345
          list: 123
      '
      to_match_snapshot

    expect
      parse_expr '{rec=123}'
      to_match_snapshot


  it 'parses unicode as keys', fn:
    expect
      parse_expr '{π, ƒ}'
      to_match_snapshot


  it 'parses member expr as key', fn:
    expect
      parse_expr '{foo.bar: 134, spam}'
      to_match_snapshot

    expect
      parse_expr '
        rec:
          foo.bar: 1234
          spam
          rec.list: 1234
      '
      to_match_snapshot

    expect
      parse_expr '{foo.bar.(spam): 1234}'
      to_match_snapshot

    expect
      parse_expr '{foo.bar=123}'
      to_match_snapshot

    expect
      parse_expr '{rec.list: lst=123, ni}'
      to_match_snapshot




describe 'records', fn:
  it 'parses', fn:
    expect
      parse_expr '
        rec:
          foo
          # prop comment
          (bar + ni): 123
          spam
          # trailing comment
      '
      to_match_snapshot



describe 'record {...} - parsing failures', fn:
  it 'errors when missing `}`', fn:
    expect
      parse '{', 'test.fnk'
      to_match_snapshot


  it 'errors when encountering bad end of prop', fn:
    expect
      parse '{foo) bar}', 'test.fnk'
      to_match_snapshot


  it 'errors when encountering bad key', fn:
    expect
      parse '{2*3: bar}', 'test.fnk'
      to_match_snapshot

    expect
      parse '{+foo}', 'test.fnk'
      to_match_snapshot

    expect
      parse '{-foo}', 'test.fnk'
      to_match_snapshot

    expect
      parse '{*foo}', 'test.fnk'
      to_match_snapshot

    expect
      parse '{/foo}', 'test.fnk'
      to_match_snapshot

    expect
      parse '{?foo}', 'test.fnk'
      to_match_snapshot

    expect
      parse '{:foo}', 'test.fnk'
      to_match_snapshot

    expect
      parse '{{foo}', 'test.fnk'
      to_match_snapshot

    expect
      parse '{}foo}', 'test.fnk'
      to_match_snapshot

    expect
      parse '{[foo}', 'test.fnk'
      to_match_snapshot

    expect
      parse '{]foo}', 'test.fnk'
      to_match_snapshot

    expect
      parse '{)foo}', 'test.fnk'
      to_match_snapshot

    expect
      parse '{,foo}', 'test.fnk'
      to_match_snapshot


  it 'errors when encountering bad start of prop key', fn:
    expect
      parse '{foo, [bar]: spam}', 'test.fnk'
      to_match_snapshot
