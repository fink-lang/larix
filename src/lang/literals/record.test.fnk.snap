// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`record {...} - parsing failures errors when encountering bad end of prop 1`] = `
module (1:0-1:4)
  rec (1:0-1:4)
    rec:kv (1:1-1:4)
      ident (1:1-1:4) foo

error:
test.fnk:1:4
1| {foo) bar}
       ^

Expected \`,\` or indented(>=1) new line or \`}\`.



`;

exports[`record {...} - parsing failures errors when encountering bad key 1`] = `
module (1:0-1:1)
  rec (1:0-1:1)
    ::error::

error:
test.fnk:1:1
1| {2*3: bar}
    ^

Expected prop-key but found '2'.



`;

exports[`record {...} - parsing failures errors when encountering bad key 2`] = `
module (1:0-1:1)
  rec (1:0-1:1)
    ::error::

error:
test.fnk:1:1
1| {+foo}
    ^

Expected prop-key but found '+'.



`;

exports[`record {...} - parsing failures errors when encountering bad key 3`] = `
module (1:0-1:1)
  rec (1:0-1:1)
    ::error::

error:
test.fnk:1:1
1| {-foo}
    ^

Expected prop-key but found '-'.



`;

exports[`record {...} - parsing failures errors when encountering bad key 4`] = `
module (1:0-1:1)
  rec (1:0-1:1)
    ::error::

error:
test.fnk:1:1
1| {*foo}
    ^

Expected prop-key but found '*'.



`;

exports[`record {...} - parsing failures errors when encountering bad key 5`] = `
module (1:0-1:1)
  rec (1:0-1:1)
    ::error::

error:
test.fnk:1:1
1| {/foo}
    ^

Expected prop-key but found '/'.



`;

exports[`record {...} - parsing failures errors when encountering bad key 6`] = `
module (1:0-1:1)
  rec (1:0-1:1)
    ::error::

error:
test.fnk:1:1
1| {?foo}
    ^

Expected prop-key but found '?'.



`;

exports[`record {...} - parsing failures errors when encountering bad key 7`] = `
module (1:0-1:1)
  rec (1:0-1:1)


error:
test.fnk:1:1
1| {:foo}
    ^

Expected \`,\` or indented(>=1) new line or \`}\`.



`;

exports[`record {...} - parsing failures errors when encountering bad key 8`] = `
module (1:0-1:1)
  rec (1:0-1:1)
    ::error::

error:
test.fnk:1:1
1| {{foo}
    ^

Expected prop-key but found '{'.



`;

exports[`record {...} - parsing failures errors when encountering bad key 9`] = `
module (1:0-1:2)
  rec (1:0-1:2)


error:
test.fnk:1:2
1| {}foo}
     ^

Expected \`,\` or indented new line or end.



`;

exports[`record {...} - parsing failures errors when encountering bad key 10`] = `
module (1:0-1:1)
  rec (1:0-1:1)
    ::error::

error:
test.fnk:1:1
1| {[foo}
    ^

Expected prop-key but found '['.



`;

exports[`record {...} - parsing failures errors when encountering bad key 11`] = `
module (1:0-1:1)
  rec (1:0-1:1)


error:
test.fnk:1:1
1| {]foo}
    ^

Expected \`,\` or indented(>=1) new line or \`}\`.



`;

exports[`record {...} - parsing failures errors when encountering bad key 12`] = `
module (1:0-1:1)
  rec (1:0-1:1)


error:
test.fnk:1:1
1| {)foo}
    ^

Expected \`,\` or indented(>=1) new line or \`}\`.



`;

exports[`record {...} - parsing failures errors when encountering bad key 13`] = `
module (1:0-1:6)
  rec (1:0-1:6)
    empty (1:2-1:2)
    rec:kv (1:2-1:5)
      ident (1:2-1:5) foo
`;

exports[`record {...} - parsing failures errors when encountering bad start of prop key 1`] = `
module (1:0-1:5)
  rec (1:0-1:5)
    rec:kv (1:1-1:4)
      ident (1:1-1:4) foo
    ::error::

error:
test.fnk:1:6
1| {foo, [bar]: spam}
         ^

Expected prop-key but found '['.



`;

exports[`record {...} - parsing failures errors when missing \`}\` 1`] = `
module (1:0-1:1)
  rec (1:0-1:1)


error:
test.fnk:1:1
1| {
    ^

Expected \`,\` or indented(>=1) new line or \`}\`.



`;

exports[`record {...} parses calculated prop 1`] = `
rec (1:0-1:28)
  rec:kv (1:1-1:4)
    ident (1:1-1:4) foo
  rec:kv (1:6-1:21)
    group (1:6-1:16)
      arithm + (1:7-1:15)
        ident (1:7-1:10) bar
        ident (1:13-1:15) ni
    number (1:18-1:21) 123
  rec:kv (1:23-1:27)
    ident (1:23-1:27) spam
`;

exports[`record {...} parses default assignment prop 1`] = `
rec (1:0-1:9)
  rec:kv (1:1-1:8)
    ident (1:1-1:4) foo
    assign = (1:1-1:8)
      ident (1:1-1:4) foo
      number (1:5-1:8) 123
`;

exports[`record {...} parses empty: {} 1`] = `
rec (1:0-1:2)

`;

exports[`record {...} parses keywords as keys 1`] = `
rec (1:0-1:60)
  rec:kv (1:1-1:3)
    ident (1:1-1:3) fn
  rec:kv (1:5-1:10)
    ident (1:5-1:10) match
  rec:kv (1:12-1:16)
    ident (1:12-1:16) fold
  rec:kv (1:18-1:24)
    ident (1:18-1:24) unfold
  rec:kv (1:26-1:29)
    ident (1:26-1:29) map
  rec:kv (1:31-1:35)
    ident (1:31-1:35) find
  rec:kv (1:37-1:43)
    ident (1:37-1:43) filter
  rec:kv (1:45-1:50)
    ident (1:45-1:50) await
  rec:kv (1:52-1:59)
    ident (1:52-1:59) attempt
`;

exports[`record {...} parses multiple shorthand exprs 1`] = `
rec (1:0-1:13)
  rec:kv (1:1-1:4)
    ident (1:1-1:4) foo
  rec:kv (1:6-1:9)
    ident (1:6-1:9) bar
  rec:kv (1:11-1:12)
    ident (1:11-1:12) π
`;

exports[`record {...} parses multiple shorthand exprs 2`] = `
rec (1:0-1:28)
  rec:kv (1:1-1:4)
    ident (1:1-1:4) foo
  rec:kv (1:6-1:14)
    ident (1:6-1:9) bar
    number (1:11-1:14) 123
  rec:kv (1:16-1:27)
    ident (1:16-1:21) shrub
    string ' (1:23-1:27)
      'ni'
`;

exports[`record {...} parses single prop 1`] = `
rec (1:0-1:11)
  rec:kv (1:1-1:10)
    ident (1:1-1:4) foo
    ident (1:6-1:10) spam
`;

exports[`record {...} parses single shorthand prop 1`] = `
rec (1:0-1:5)
  rec:kv (1:1-1:4)
    ident (1:1-1:4) foo
`;

exports[`record {...} parses single shorthand value 1`] = `
rec (1:0-1:7)
  rec:kv (1:1-1:6)
    string ' (1:1-1:6)
      'foo'
`;

exports[`record {...} parses single str prop} 1`] = `
rec (1:0-1:13)
  rec:kv (1:1-1:12)
    string ' (1:1-1:6)
      'foo'
    ident (1:8-1:12) spam
`;

exports[`record {...} parses spread prop 1`] = `
rec (1:0-1:13)
  rec:kv (1:1-1:4)
    ident (1:1-1:4) foo
  rec:kv (1:6-1:12)
    spread ... (1:6-1:12)
      ident (1:9-1:12) bar
`;

exports[`record {...} parses unicode as keys 1`] = `
rec (1:0-1:6)
  rec:kv (1:1-1:2)
    ident (1:1-1:2) π
  rec:kv (1:4-1:5)
    ident (1:4-1:5) ƒ
`;

exports[`records parses 1`] = `
rec (1:0-7:0):
  rec:kv (2:2-2:5)
    ident (2:2-2:5) foo
  rec:kv (4:2-4:17)
    # prop comment
    group (4:2-4:12)
      arithm + (4:3-4:11)
        ident (4:3-4:6) bar
        ident (4:9-4:11) ni
    number (4:14-4:17) 123
  rec:kv (5:2-5:6)
    ident (5:2-5:6) spam
`;
